### PR TITLE
Handle browser visibility events

### PR DESCRIPTION
**Pull request by Sapphire64**
Yo. Had problemswith getting some messages when browser was in background. Visibility was stating that we were visible but we were not.

Now it works perfectly for me.

### CODE REVIEW

**Comment by Sapphire64** [comment]
@Pomax I have two more updates but first maybe we should arrange this merge? :)

File: index.js [hunk 1]
```diff
@@ -44,12 +58,13 @@
...
       if(visible) {
-        var d0 = (gcs.getPropertyValue("display") === "none");
-        var o0 = (gcs.getPropertyValue("opacity") === 0);
-        var v0 = (gcs.getPropertyValue("visibility") === "hidden");
-        visible = visible && !d0 && !o0 && !v0;
+        var isDocHidden = document[hidden];
+        var isElementNotDisplayed = (gcs.getPropertyValue("display") === "none");
+        var elementHasZeroOpacity = (gcs.getPropertyValue("opacity") === 0);
+        var isElementHidden = (gcs.getPropertyValue("visibility") === "hidden");
+        visible = visible && !isDocHidden && !isElementNotDisplayed && !elementHasZeroOpacity && !isElementHidden;
```
Comment (Pomax) [inline index.js]: minor optimization: turn the four negations into a shortcut single:

```
visible = visible && !(isDocHidden || isElementNotDisplayed || elementHasZeroOpacity || isElementHidden)
```

File: index.js [hunk 2]
```diff
@@ -1,9 +1,23 @@
 (function() {
-
   var RATE_LIMIT = 25;
 
-  var ComponentVisibilityMixin = {
+  /* Finding browser event names for hidden window event */
+  var hidden, visibilityChange;
+  if (typeof document.hidden !== "undefined") { // Opera 12.10 and Firefox 18 and later support
```
Comment (Pomax) [inline index.js]: good idea to rewrite this as a prefix array with a `some` instead:

```
['',"moz","ms","webkit"].some(function(prefix) {
  if (typeof document[prefix + "hidden"] !== undefined) {
    hidden = prefix + "hidden";
    visibilityChange = prefix + "visibilitychange";
    return true;
  }
});
```

That said, the target here is modern HTML5 browsers. If the latest versions of Firefox and Chrome, and IE 10 and 11, support this without prefixes, let's take out the prefixes.

```
```

File: index.js [hunk 3]
```diff
@@ -89,6 +104,7 @@
           }
         }.bind(this), RATE_LIMIT);
       }.bind(this);
+      document.addEventListener(visibilityChange, this._rcv_fn, false);
```
Comment (Pomax) [inline index.js]: no need to add the third argument _unless_ it's `true` (which is an absolutely useless flag anyway).

File: index.js [hunk 4]
```diff
@@ -44,12 +58,13 @@
           visible = horizontallyVisible && verticallyVisible;
 
       // but let's be fair: if we're opacity: 0 or
-      // visibility: hidden, we're not visible at all.
+      // visibility: hidden, or browser window is closed we're not visible at all.
```
Comment (Pomax) [inline index.js]: I'd say "minimised", because closed generally means you exited the program.

**Comment by Sapphire64** [comment]
@Pomax I've fixed according to your suggestions (by rewriting that commit). 

```
If the latest versions of Firefox and Chrome, and IE 10 and 11, support this without prefixes, let's take out the prefixes.
```

Unfortunately I have no ways to test in IE, I've tried my stuff only in Chrome.

File: bower.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "react-component-visibility",
   "main": "index.js",
-  "version": "0.0.5",
+  "version": "0.0.6",
```
Comment (Pomax) [inline bower.json]: no need to pump this until a 0.0.6 tag actually exists on github, which I do when I release a new version, so let's leave this as 0.0.5 (also, bower technically ignoes the version _entirely_)

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-component-visibility",
-  "version": "0.0.5",
+  "version": "0.0.6",
```
Comment (Pomax) [inline package.json]: this needs to stay 0.0.5, `npm version patch` will automatically update this

File: index.js [hunk 7]
```diff
@@ -1,9 +1,16 @@
 (function() {
-
   var RATE_LIMIT = 25;
 
+  /* Finding browser event names for hidden window event */
+  var hidden, visibilityChange;
+  ['',"moz","ms","webkit"].some(function(prefix) {
```
Comment (Pomax) [inline index.js]: tested this in FF, Chrome and IE11, and all of them work without prefix, so let's remove this and just use the proper events and state names in the code

File: index.js [hunk 8]
```diff
@@ -89,6 +99,7 @@
           }
         }.bind(this), RATE_LIMIT);
       }.bind(this);
+      document.addEventListener(visibilityChange, this._rcv_fn);
```
Comment (Pomax) [inline index.js]: `addEventListener("visibilitychange", this._rcv_fn);`

File: index.js [hunk 9]
```diff
@@ -44,12 +51,15 @@
...
 
       // but let's be fair: if we're opacity: 0 or
-      // visibility: hidden, we're not visible at all.
+      // visibility: hidden, or browser window is minimized we're not visible at all.
       if(visible) {
-        var d0 = (gcs.getPropertyValue("display") === "none");
-        var o0 = (gcs.getPropertyValue("opacity") === 0);
-        var v0 = (gcs.getPropertyValue("visibility") === "hidden");
-        visible = visible && !d0 && !o0 && !v0;
+        var isDocHidden = document[hidden];
```
Comment (Pomax) [inline index.js]: var isDocHidden = document.hidden;

**Comment by Pomax** [comment]
tests show the proper API calls work in all modern HTML5 browsers, so let's go with just writing the code on that assumption. We can take out the prefix shim, and that simplifies the code nicely. The code works well in FF, Chrome and IE11: window minimize/restore shows visible=>invisible=>visible changes occuring, so that part looks great.

**Comment by Sapphire64** [comment]
hey, thank you very much for review! I'm having not much time today and hopefully I'll have time to fix my PR on weekends. If not - on Monday :) have a great weekends!

**Comment by Pomax** [comment]
coolios, looking forward to it!

**Comment by Sapphire64** [comment]
@Pomax it should be much better now. Please let me know if you have more suggestions.

**Comment by Pomax** [comment]
nope, this is a sweet improvement, thanks so much for the PR! (and curious what additional improvements you were thinking of after this lands =)

**Pull request merged**